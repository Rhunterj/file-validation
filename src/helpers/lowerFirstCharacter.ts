export const lowerFirstCharacter = (str: string) => {
  return str[0].toLowerCase() + str.slice(1);
}