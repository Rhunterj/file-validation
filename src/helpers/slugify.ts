export const slugify = (str: string): string => {
  if (!str) {
    return str
  }

  return str.toLowerCase().replace(/[^a-zA-Z0-9]+(.)/g, (m, chr) => chr.toUpperCase());
}