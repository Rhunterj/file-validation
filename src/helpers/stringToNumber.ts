export const convertStringToNumber = (str: string) => {
  const number = Number(str);

  if (isNaN(number)) {
    return str;
  }

  return number;
}