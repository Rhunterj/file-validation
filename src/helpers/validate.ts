import { convertedData } from "../types";

export const validate = (obj: convertedData[]) => {
  const errors: any = [];
  const hasDuplicateReference = obj.reduce((acc, curr) => {
    acc[curr.reference] = ++acc[curr.reference] || 0;
    return acc
  }, {} as Record<string, number>)

  const duplicateReferences = obj.filter(item => hasDuplicateReference[item.reference]);
  errors.duplicateKeys = duplicateReferences;

  const checkEndbalance = (item: convertedData) => item.endBalance !== Math.round((item.startBalance + item.mutation) * 100) / 100;
  const hasInvalidEndBalance = obj.filter((item) => checkEndbalance(item));

  errors.invalidEndBalance = hasInvalidEndBalance;

  return errors;
}