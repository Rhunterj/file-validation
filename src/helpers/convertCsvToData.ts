import { convertedDataType } from "../types";
import { slugify } from "./slugify";
import { convertStringToNumber } from "./stringToNumber";

export const convertCsvToData = (csv: string): convertedDataType[] => {
  const lines = csv.split("\n").slice(1);
  lines.pop(); // remove last empty line

  return lines.map(line => {
    const values = line.split(",");
    const headers = csv.split("\n")[0].split(",");

    return headers.reduce((obj: any, header: string, index: number) => {
      obj[slugify(header)] = convertStringToNumber(values[index]);
      return obj;
    }, {}) as convertedDataType
  });
}