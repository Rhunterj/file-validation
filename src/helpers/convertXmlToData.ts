import { convertedDataType } from "../types";
import { convertStringToNumber } from "./stringToNumber";

export const convertXmlToData = (xml: string): convertedDataType[] => {
  const parser = new DOMParser();
  const xmlDoc = parser.parseFromString(xml, "text/xml");
  const records = xmlDoc.querySelectorAll('record');

  return Array.from(records, record => {
    const obj: Record<string, any> = {};
    obj['reference'] = Number(record.getAttribute("reference"));
    Array.from(record.children).forEach(child => {
      obj[child.tagName] = convertStringToNumber(child.innerHTML);
    });
    return obj as convertedDataType;
  });
}