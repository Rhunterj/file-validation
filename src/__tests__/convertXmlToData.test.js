import { convertXmlToData } from '../helpers/convertXmlToData';

test("convertXmlToData function", () => {
  const result = convertXmlToData('<records><record reference="123"><accountNumber>NL123</accountNumber><description>description</description><startBalance>1.23</startBalance><mutation>1.23</mutation><endBalance>1.23</endBalance></record></records>');
  
  expect(result).toEqual([{
    reference: 123,
    accountNumber: 'NL123',
    description: 'description',
    startBalance: 1.23,
    mutation: 1.23,
    endBalance: 1.23
  }]);
  expect(result).toMatchSnapshot();
});