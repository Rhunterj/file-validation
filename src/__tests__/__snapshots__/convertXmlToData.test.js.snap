// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convertXmlToData function 1`] = `
Array [
  Object {
    "accountNumber": "NL123",
    "description": "description",
    "endBalance": 1.23,
    "mutation": 1.23,
    "reference": 123,
    "startBalance": 1.23,
  },
]
`;
